### DISCUSSION TITLE
ODB:: Is there any way to unfrozen the master? (Changing the size of dbMaster during the code is running)

### DISCUSSION THREAD

**Discussion by ApeachM** [The-OpenROAD-Project/OpenROAD#3644] [Q&A] [answered]
```c++
// master setting
dbMaster *master = dbMaster::create(db_lib, lib_cell_name.c_str());
master->setWidth(width);
master->setHeight(height);
master->setType(dbMasterType::CORE);
...
master->setFrozen();
// connect inst to the master
    
dbInst* inst = dbInst::create(db_block, master, instance_info->inst_name.c_str());

// now I want to change the size of inst, by unfrozen master and resize width and height
// Is my intention of the below pseudo-code possible?
master->setUnFrozen();
master->setWidth(new_width);
master->setHeight(new_height);
master->setFrozen();
```

**I would like to know whether the above pseudo-code is possible.**

**I want to resize the cell size** after the cell size is already determined. (For example, after parsing lef and def).
But it seems like not to exist function for unfreezing.

If I want to change the cell size, then should I 
1. re-construct the instance,
2. re-construct the new dbMaster, 
3. deleting the original instance, 
4. and re-connecting the new dbInst to the nets, which was connected with the original one?

I would like to know whether **is there another way, which is more simple**.

Thanks for the time for reading my question.

Sincerely.

**Comment by maliberty** [comment]
Why are you freezing it in the first place?

**Reply by ApeachM** [reply]
Thank you for the reply.

I want to change the cell size after parsing lef and def.
But as far as I know, as the necessary condition for connecting dbInst and dbMaster, the dbMaster should be frozen before that.
I just checked that the dbMaster was frozen, by calling the below lines of code.
I checked for several dbMaster, but they were all frozen.
```c++
  odb::lefin lef_reader(db_database, &logger_, false);
  odb::dbLib *lib = lef_reader.createTechAndLib("lib_name", lef_file.c_str()); 

  odb::defin def_reader(db_database, &logger_);
  vector<odb::dbLib *> search_libs;
  for (odb::dbLib *lib : db_database->getLibs())
    search_libs.push_back(lib);
  odb::dbChip *chip = def_reader.createChip(search_libs, def_file.c_str());


  dbMaster* db_master = bench_information_normal_->findMaster("NAND3X8");
  bool is_frozen = db_master->isFrozen();
```

This is the reason why I consider the state of dbMaster, after frozen.

Thanks, sincerely.

**Comment by maliberty** [comment, marked as answer]
setWidth & setHeight don't care about the frozen status afaik.  The key thing is that the port list is frozen.  You can't add a dbMTerm to a frozen master.
